eth-contracts/contracts/CKBChain.sol

```diff
@@ -60,28 +71,233 @@ contract CKBChain is ICKBChain, ICKBSpv {
 
-    /// #ICKBChain
+    /**
+     * @dev Throws if called by any account other than the governance.
+     */
+    modifier onlyGov() {
+        require(msg.sender == governance, "caller is not the governance");
+        _;
+    }
+
+    constructor() public {
+        governance = msg.sender;
+    }
+
+    // query
+    function getLatestBlockNumber() public returns (uint64) {
+        return latestBlockNumber;
+    }
+
+    // query
+    function getHeadersByNumber(uint64 blockNumber) public returns (bytes32[] memory) {
+        return allHeaderHashes[blockNumber];
+    }
+
+    // query
+    function getCanonicalHeaderHash(uint64  blockNumber) public returns (bytes32) {
+        return canonicalHeaderHashes[blockNumber];
+    }
+
+    // query
+    function getCanonicalTransactionsRoot(bytes32 blockHash) public returns (bytes32) {
+        return canonicalTransactionsRoots[blockHash];
+    }
+
+    // query
+    function getLatestEpoch() public returns (uint64) {
+        return latestHeader.epoch;
+    }
+
+    function initWithHeader(bytes calldata data, bytes32 blockHash, uint64 finalizedGcThreshold, uint64 canonicalGcThreshold) external onlyGov {
+        require(!initialized, "Contract is already initialized");
+        initialized = true;
+
+        // set init threshold
+        FinalizedGcThreshold = finalizedGcThreshold;
+        CanonicalGcThreshold = canonicalGcThreshold;
+
+        // decoder init header
+        bytes29 rawHeaderView = data.ref(uint40(ViewCKB.CKBTypes.Header)).rawHeader();
+        (uint256 target,) = CKBPow.compactToTarget(rawHeaderView.compactTarget());
+        uint256 difficulty = CKBPow.targetToDifficulty(target);
+        uint64 blockNumber = rawHeaderView.blockNumber();
+        initBlockNumber = blockNumber;
+        BlockHeader memory header = BlockHeader(
+            blockNumber,
+            rawHeaderView.epoch(),
+            difficulty,
+            difficulty,
+            rawHeaderView.parentHash()
+        );
+
+        // set headers
+        allHeaderHashes[blockNumber].push(blockHash);
+        blockHeaders[blockHash] = header;
+
+        // set canonical chain
+        _refreshCanonicalChain(header, blockHash);
+        canonicalTransactionsRoots[blockHash] = rawHeaderView.transactionsRoot();
+    }
+
+    /// # ICKBChain
+    // TODO add onlyGov for phase1 production
     function addHeaders(bytes calldata data) external {
-        // TODO addHeaders
+        require(initialized, "Contract is not initialized");
+
+        // 1. view decode from data to headers view
+        bytes29 headerVecView = data.ref(uint40(ViewCKB.CKBTypes.HeaderVec));
+
+        // 2. iter headers
+        uint32 length = headerVecView.lengthHeaderVec();
+        uint32 index = 0;
+        while (index < length) {
+            bytes29 headerView = headerVecView.indexHeaderVec(index);
+            _addHeader(headerView);
+            index++;
+        }
+    }
+
+    function _addHeader(bytes29 headerView) private {
```

Comment:
why so many verify header code  if we use owner mod temporarily, we can just maintain best chain according to total difficulty

Reply:
We might dev optimistic plan base on the phase1 code, so there's no need to delete this code if it doesn't make a negative impact